### PR TITLE
Improve FLUX image-to-image (Trajectory Guidance)

### PR DESCRIPTION
**Pull request by RyanJDick**
## Summary

This PR makes some improvements to the FLUX image-to-image and inpainting behaviours.

Changes:
- Expand inpainting region at a cutoff timestep. This improves seam coherence around inpainting regions.
- Add Trajectory Guidance to improve the ability to control how much an image gets modified during image-to-image/inpainting (see the code for a more technical explanation - it's well-documented).

## `trajectory_guidance_strength` Usage

- The `trajectory_guidance_strength` param has been added to the `FLUX Denoise` invocation.
- `trajectory_guidance_strength` defaults to `0.0` and should be in the range [0, 1].
- `trajectory_guidance_strength = 0.0` has no effect on the denoising process.
- `trajectory_guidance_strength = 1.0` will guide strongly towards the original image.

## FLUX image-to-image usage tips

- As always, prompt matters a lot.
- If you are trying to making minor perturbations to an image, use vanilla image-to-image by setting the `denoising_start` param.
- If you are trying to make significant changes to an image, using trajectory guidance will give more control than using vanilla image-to-image. Set `denoising_start=0.0` and adjust `trajectory_guidance_strength` to control the amount of change in the image.
- The 'transition point' where the image changes the most as you adjust `trajectory_guidance_strength` or `denoise_start` varies depending on the noise. So, set a fixed noise seed, then tune those params.

## QA Instructions

- [x] Vanilla image-to-image - No change in output
- [x] Vanilla inpainting - No change in output
- [x] Vanilla outpainting - No change in output
- Trajectory Guidance image-to-image
    - [x] TGS = 0.0 is identical to Vanilla case
    - [x] TGS = 1.0 guides close to the original image
      - Not as close as I'd like, but it's not broken.
    - [x] Smooth transition as TGS varies
    - [x] Smoke test: TGS with denoise_start > 0.0
- TG inpainting
    - [x] TGS = 0.0 is identical to Vanilla case
    - [x] TGS = 1.0 guides close to the original image
      - Not as close as I'd like, but it's not broken
    - [x] Smooth transition as TGS varies
    - [x] Smoke test: TGS with denoise_start > 0.0
- TG outpainting
    - [x] TGS = 0.0 is identical to Vanilla case
    - [x] Smoke test TGS outpainting
- [x] Smoke test FLUX text-to-image
- [x] Preview images look ok for all of above.

## Known issues (will be addressed in follow-up PRs)

- The current TGS scale biases towards creating more change than desired in the image. More tuning of the TG change schedule is required.
- TGS does not work very well for outpainting right now. This _might_ be solvable, but more likely we'll just want to discourage it in the Linear UI.

## Merge Plan

No special instructions.

## Checklist

- [x] _The PR has a short but descriptive title, suitable for a changelog_
- [x] _Tests added / updated (if applicable)_
- [x] _Documentation added / updated (if applicable)_

### CODE REVIEW

File: invokeai/app/invocations/flux_denoise.py [hunk 1]
```diff
@@ -68,6 +68,12 @@ class FluxDenoiseInvocation(BaseInvocation, WithMetadata, WithBoard):
         description=FieldDescriptions.denoising_start,
     )
     denoising_end: float = InputField(default=1.0, ge=0, le=1, description=FieldDescriptions.denoising_end)
+    trajectory_guidance_strength: float = InputField(
+        default=0.0,
+        ge=0.0,
+        le=1.0,
+        description="Value indicating how strongly to guide the denoising process towards the initial latents (during image-to-image). Range [0, 1]. A value of 0.0 is equivalent to vanilla image-to-image. A value of 1.0 will guide the denoising process very close to the original latents.",
+    )
```
[the remark this answers is not in the document]
Comment (brandonrising) [reply invokeai/app/invocations/flux_denoise.py:76]: This makes sense to me. I'd prefer this to adding a selector

File: invokeai/backend/flux/trajectory_guidance_extension.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        change_ratio_at_t_1 = build_line(x1=0.0, y1=1.0, x2=1.0, y2=0.0)(self._trajectory_guidance_strength)
+        change_ratio_at_cutoff = 1.0
+        t_cutoff = build_line(x1=0.0, y1=1.0, x2=1.0, y2=0.5)(self._trajectory_guidance_strength)
+        change_ratio = 1.0
+        if t_prev > t_cutoff:
+            # If we are before the cutoff, linearly interpolate between the change_ratio at t=1.0 and the change_ratio
+            # at the cutoff.
+            change_ratio = build_line(x1=1.0, y1=change_ratio_at_t_1, x2=t_cutoff, y2=change_ratio_at_cutoff)(t_prev)
+
+        mask = mask * change_ratio
```
Comment (brandonrising) [inline invokeai/backend/flux/trajectory_guidance_extension.py:95]: Does this need to be clamped

File: invokeai/backend/flux/trajectory_guidance_extension.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    def step(
+        self, t_curr_latents: torch.Tensor, pred_noise: torch.Tensor, t_curr: float, t_prev: float
+    ) -> torch.Tensor:
+        # Handle gradient cutoff.
+        mask = self._apply_mask_gradient_adjustment(t_prev)
+
+        # Calculate the change_ratio based on the trajectory_guidance_strength.
+        # These mappings from trajectory_guidance_strength have no theoretical basis - they were tuned manually.
+        change_ratio_at_t_1 = build_line(x1=0.0, y1=1.0, x2=1.0, y2=0.0)(self._trajectory_guidance_strength)
```
[the remark this answers is not in the document]
Comment (RyanJDick) [reply invokeai/backend/flux/trajectory_guidance_extension.py:86]: The performance benefit will probably be negligible, but might as well pre-compute. I'll address this.

File: invokeai/frontend/web/src/features/nodes/util/graph/generation/buildFLUXGraph.ts [hunk 4]
```diff
@@ -125,9 +138,15 @@ export const buildFLUXGraph = async (
...
-      1 - img2imgStrength,
+      denoisingStart,
       false
     );
+    if (optimizedDenoisingEnabled) {
+      g.updateNode(noise, {
+        denoising_start: 0,
+        trajectory_guidance_strength: img2imgStrength,
+      });
+    }
```
Comment (RyanJDick) [inline invokeai/frontend/web/src/features/nodes/util/graph/generation/buildFLUXGraph.ts:149]: We did this in a few places:
```suggestion
    if (optimizedDenoisingEnabled) {
      g.updateNode(noise, {
        denoising_start: 0,
        trajectory_guidance_strength: 1 - img2imgStrength,
      });
    }
```

File: invokeai/backend/flux/trajectory_guidance_extension.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (RyanJDick) [reply invokeai/backend/flux/trajectory_guidance_extension.py]: [16ca540](https://github.com/invoke-ai/InvokeAI/pull/6900/commits/16ca540ecedbf857160dc7fab52f7bdeccf3a578)

File: invokeai/backend/flux/trajectory_guidance_extension.py (hunk 2 above)
Comment (RyanJDick) [reply invokeai/backend/flux/trajectory_guidance_extension.py]: It shouldn't need to be clamped, but I've added an assertion to verify that it is in the expected range so we can't screw it up: [cd3a7bd](https://github.com/invoke-ai/InvokeAI/pull/6900/commits/cd3a7bdb5ee3a37265e7c426bd9fbc88d95c5863)

**Review by hipsterusername — APPROVED** [review]

**Pull request merged**